### ISSUE TITLE
User Profiles

### ISSUE THREAD

**Issue by DatNightcat** [blobs-io/blobs.io#80]
All users would automatically get a user profile.

User profiles would show:
- Noms (per gamemode and total)
- Deaths (per gamemode and total)
- Nom:Death ratio (per gamemode)
- Total time spent in-game (per gamemode and total)
- Pixels travelled (per gamemode and total)
- Time:Pixel ratio (per gamemode) (you don’t have to add this)
- Exact time joined
- Time remaining on ban
- Bio (optional)
- Birthday (optional)
- Blob Badges

Blob Badges would display next to the username, could show a variety of achievements, and show how the user got them upon hovering over the badge. I thought of a LOT of badges. Here’s what I came up with:
- Developer Blob — This user is currently recognized as a developer.*
- Staff Blob — This user moderates blobs.io in some form.*
- Contributor Blob — This user has contributed to the game in some form.+>
  - Coder Blob — This user has contributed code to the game.+>
    - Cascading Blob — This user has contributed CSS to the game.+
    - Bug Blaster Blob — This user has contributed a bug fix to the game.+
  - Designer Blob — This user has contributed artwork that has appeared in-game at some point.+
  - Bug Catcher Blob — This user has reported a bug that was confirmed.+
  - Enchancer Blob — This user has submitted a feature request that became part of the game in some form.+
- Verified Blob — This user has been confirmed to be who they say they are.*
- Alpha Blob — This user signed up before the game even had multiplayer!
- Beta Blob — This user has been confirmed to have participated in the beta testing period of the game.
- Yearly Blob — This user’s account was created 1 year ago.+
- Party Blob — It’s this user’s birthday!*
- Clean Blob — This user has had no punishments in the past 3 months.*+

An asterisk indicates that the Blob Badge could be revoked from the user, either automatically or manually.
A plus sign indicates that you could get as many of the Blob Badge as you wanted.
A greater than sign indicates that this Blob Badge has sub-badges. If you hovered over the parent Blob Badge, the sub-badges would temporarily slide out.

How Blob Badges marked with a plus sign would work:
- Contributor Blob Badges — Given out per pull request, bug report, or feature request. If the contribution is too small, users may need to contribute more in order to get a badge. However, it wouldn’t have to be all at once.
- Yearly Blob Badges — Given out on the account’s yearly anniversary.
- Clean Blob Badges — Given out every 3 months that the account has had no punishments. If the user is punished, **they would lose ALL of their Clean Blob Badges!**

I didn’t think of any gameplay badges because the game doesn’t have too much to offer right now.

**Comment by y21** [comment]
Looks good to me. I approve. But:

> * Nom:Death ratio (per gamemode)

I assume "nom" in this case means how often someone "won" against another blob?

Right now the `accounts` table in the database stores the following:
- Username
- **Hashed** password (not plain-text)
- Registered at (timestamp)
- Blob Rating (BR)
- Role (Admin, Guest or normal account)
- Blobcoins
- Timestamp from when the user has claimed their daily gift the last time (used to determine how long the user needs to wait until they can claim it again)
- Travelled distance
- Blobs
- Current blob (what blob the user is currently using)
- Clan (__unused value__; see #29)

In other words: right now it does not store everything you suggested and we would need to add columns to the table, but that should not be a problem.


> * Alpha Blob — This user signed up before the game even had multiplayer!

I am unsure if we can limit it to users who have registered *before* the game had multiplayer because we started logging the registration date kind of late. Perhaps we can limit it to users who have registered in 2018?


But other than that, it looks like something we could add.

**Comment by DatNightcat** [comment]
> I assume "nom" in this case means how often someone "won" against another blob?

Yes.
> I am unsure if we can limit it to users who have registered _before_ the game had multiplayer because we started logging the registration date kind of late. Perhaps we can limit it to users who have registered in 2018?

I was going to leave the cutoff date for being able to obtain this Blob Badge up for discussion, but I just spat that out. I’m pretty sure that you couldn’t even sign up at that time!

Looking at the commit history, I think a suitable cutoff date would be October 13th, 2018, but you should probably look into how many people would get the Blob Badge, because we wouldn’t want only cool people that we know to have the badge. Assuming how long the game stays in beta period, we might even need to move the cutoff date into 2019!

**Comment by y21** [comment]
If we were to set the date to October 13th, only you and me would get it. Most of the users have a registration date of November 1st (12 out of 33), because we had to set everyones' date to a dummy value as it was unclear when exactly they made their account.

**Comment by DatNightcat** [comment]
> If we were to set the date to October 13th, only you and me would get it.

1. That’s sad.
2. Alright. Let’s set the cutoff date to the first account created after November 1st.
